### ISSUE TITLE
fix(solvers): `Mask2D` is cast to float32 in `_resolve_mask_arr` / `SolveDomain`, silently downcasting weakly-typed float64 fields through the whole CG solve (stalls and raises at moderate |b|)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#302]
Part of #277.

## Summary

`_resolve_mask_arr` (`finitevolx/_src/solvers/elliptic.py:214`) and `SolveDomain` (`inhomogeneous.py:97`) convert a `Mask2D` to a float array with `dtype=jnp.float32`. Any field built from `jnp.arange(...) * dx` with Python scalars (which is what every docs notebook does) is a **weakly-typed** float64 array; multiplying it by a strongly-typed float32 mask yields float32, and `masked_laplacian` (which correctly uses `psi.dtype`) then runs the entire CG iteration in single precision under x64.

## Reproduction

```python
import jax; jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp, numpy as np
from finitevolx import CartesianGrid2D, Mask2D, streamfunction_from_vorticity
N = 34; g = CartesianGrid2D.from_interior(N-2, N-2, Lx=1.0, Ly=1.0)
m = np.zeros((N, N), bool); m[1:-1, 1:-1] = True; mask = Mask2D.from_mask(jnp.asarray(m))
x = jnp.arange(N) * g.dx                                   # weakly-typed float64
rhs = 800.0 * jnp.sin(3 * jnp.pi * x[None, :]) * jnp.sin(2 * jnp.pi * x[:, None])
print(rhs.dtype, rhs.weak_type)                            # float64 True
psi = streamfunction_from_vorticity(rhs, g.dx, g.dy, lambda_=4.0, method="cg", mask=mask)
print(psi.dtype)                                           # float32 today
```

```
field from arange + python scalars: dtype float64 weak_type True -> * float32 mask -> float32
masked_laplacian(weak, m32).dtype = float32
lambda=4.0 weak rhs + Mask2D:   out dtype = float32     <- silently single precision
|b| ≈ 790 at 34²: Mask2D path RAISES EquinoxRuntimeError (max_steps; lineax CG stalls at |r| ≈ 1e-2)
                  float64 mask converges in 10 steps
```

## Proposed fix

`jnp.asarray(mask.xy_corner_strict, dtype=rhs.dtype)` in `_resolve_mask_arr` (take the rhs dtype as an argument), and keep the mask as `bool` in `SolveDomain` (use `jnp.where` instead of multiplication). Grep for other `dtype=jnp.float32` on masks (`elliptic.py:214`, `inhomogeneous.py:97` are the two found).

## Acceptance criteria

- [ ] Test: weakly-typed float64 rhs + `Mask2D` → float64 output and convergence in ≤ 15 iterations at 34² with |b| ≈ 800.
- [ ] Test: float32 rhs stays float32.
